typescript/src/services/redemptions/redemptions-service.ts

```diff
@@ -170,6 +184,94 @@ export class RedemptionsService {
+   *                on-chain token contract.
+   * @param potentialCandidateWallets Array of wallets that can handle the
+   *                                  redemption request. The wallets must
+   *                                  be in the Live state.
+   * @returns Object containing:
+   *          - Bitcoin public key of the wallet asked to handle the redemption.
+   *           Presented in the compressed form (33 bytes long with 02 or 03 prefix).
+   *         - Wallet public key hash.
+   *         - Main UTXO of the wallet.
+   *         - Redeemer output script.
+   *
+   * @throws Throws an error if no valid redemption wallet exists for the given
+   *         input parameters.
+   */
+  protected async determineValidRedemptionWallet(
+    bitcoinRedeemerAddress: string,
+    amount: BigNumber,
+    potentialCandidateWallets: Array<SerializableWallet>
+  ): Promise<{
+    walletPublicKey: Hex
+    mainUtxo: BitcoinUtxo
+    redeemerOutputScript: Hex
+  }> {
+    let walletPublicKey: Hex | undefined = undefined
+    let mainUtxo: BitcoinUtxo | undefined = undefined
+    const redeemerOutputScript = await this.getRedeemerOutputScript(
+      bitcoinRedeemerAddress
+    )
+
+    for (let index = 0; index < potentialCandidateWallets.length; index++) {
+      const serializableWallet = potentialCandidateWallets[index]
+      const {
+        walletBTCBalance: candidateBTCBalance,
+        walletPublicKey: candidatePublicKey,
+        mainUtxo: candidateMainUtxo,
+      } = this.fromSerializableWallet(serializableWallet)
+
+      if (candidateBTCBalance.lt(amount)) {
+        console.debug(
+          `The wallet (${candidatePublicKey.toString()})` +
+            `cannot handle the redemption request. ` +
+            `Continue the loop execution to the next wallet...`
+        )
+        continue
+      }
+
+      const pendingRedemption =
+        await this.tbtcContracts.bridge.pendingRedemptions(
+          candidatePublicKey,
+          redeemerOutputScript
+        )
+
+      if (pendingRedemption.requestedAt !== 0) {
+        console.debug(
+          `There is a pending redemption request from this wallet to the ` +
+            `same Bitcoin address. Given wallet public key` +
+            `(${candidatePublicKey.toString()}) and redeemer output script ` +
+            `(${redeemerOutputScript.toString()}) pair can be used for only one ` +
+            `pending request at the same time. ` +
+            `Continue the loop execution to the next wallet...`
+        )
+        continue
+      }
+      walletPublicKey = candidatePublicKey
+      mainUtxo = candidateMainUtxo
+
+      console.debug(
+        `The wallet (${walletPublicKey.toString()})` +
+          `can handle the redemption request. ` +
+          `Stop the loop execution and proceed with the redemption...`
+      )
+
+      break
+    }
+
+    if (!walletPublicKey || !mainUtxo) {
+      throw new Error(`Could not find a wallet with enough funds.`)
+    }
```

Comment:
Nit:

```suggestion
    if (!walletPublicKey || !mainUtxo) {
      return null;
    }
```

Reply:
I believe it’s important that this error is thrown, because `requestRedemption` triest to fetch the data, determine the valid redemption wallet from that data before it can proceed with the transaction. If any of those steps throws an error, the method should fall back to the active search (performed on the client) to locate and validate the wallet.

Therefore, we want the failure to surface, to be specified so we can track it, and move to the fallback.

https://github.com/keep-network/tbtc-v2/blob/0528fcd08b28a2436e507400903431a6271e66ab/typescript/src/services/redemptions/redemptions-service.ts#L55-L103